### DISCUSSION TITLE
Performance question on @Reference

### DISCUSSION THREAD

**Discussion by ball3t** [MorphiaOrg/morphia#2825] [Q&A]
Hello,

I have an collection with an list of embedded doc with a reference to another collection. 
e.g. Collection A with a list called list1, holding embedded doc with a reference to collection B (b_id), like below:

```
Collection A:
{
    ...
    list1:[
        {
        ...
        b_id: ObjectId
        },
        {
        ...
        b_id: ObjectId
        }...
    ]
}
```

It is defined with @Reference(lazy = true) which is fine for most of the time when i don't need the data in collection B. But when I need the data in collection B, it will query 1 by 1. when there are list of 100 entry in list1, there will be 100 query. Are there any way to group the 100 query into 1?

Thanks.
 
Regards

**Comment by evanchooly** [comment]
When a list of references is fetched, it's a multi-step process:
1.  group references by targeted collections (with polymorphic types, not all subtypes are necessarily mapped to the same collection)
2. execute 1 query for each collection
3. map those entities back to the IDs/positions in the original list

So there's a little bit of work to process them all but the number of queries is minimized as much as possible.  Hope this helps.

**Reply by ball3t** [reply]
Thanks for the advice, however I have further questions. 
I tried to add the "batch" query in my application, however when I tried to call the functions inside Object B, it is already calling the query even it is defined as Lazy.
i.e. I was trying to get a list of id of B in an record of collection A.

    List<ObjectId> b_ids = new ArrayList<>();
    a.getBs().forEach(b->b_ids.add(b.get_id()));

Could you advise on how I can get the list of ids?

**Reply by ball3t** [reply]
Hello @evanchooly ,

I have used another way to solve this issue. However, i would like to see if this is a possible update for Morphia:

Current situation: When working with @Reference(lazy = true). , morphia automatically generated a query to the db to get the whole object when it calls any get methods.
Possible change: If calling the getter of the referenced field, ie _id, it won't trigger another query.


Regards

**Reply by evanchooly** [reply]
Certain methods are intercepted before a fetch is issued.  Use `@IdGetter` on the get for your ID to prevent the unnecessary fetch.

see https://github.com/MorphiaOrg/morphia/blob/master/core/src/main/java/dev/morphia/mapping/codec/references/ReferenceProxy.java#L32
